cli/functionary/client.py

```diff
@@ -0,0 +1,104 @@
+import json
+
+import click
+import requests
+
+from .config import get_config_value
+
+
+def get(endpoint):
+    """
+    Gets any data associated with an endpoint from the api
+
+    Args:
+        Endpoint: the name of the endpoint to get data from
+
+    Returns:
+        Data: Data from endpoint as Python list/dict
+
+    """
+    response = _send_request(endpoint, "get")
+    data = json.loads(response.text).get("results")
+    return data
+
+
+def post(endpoint, data=None, files=None):
+    """
+    Post provides data or files to endpoint
+
+    Args:
+        Endpoint: the name of the endpoint to get data from
+        Data: Any data to put in the request's data field
+        Files: Any files to put in the request's files field
+
+    Returns:
+        Data: Response from endpoint as Python list/dict
+
+    """
+    response = _send_request(endpoint, "post", post_data=data, post_files=files)
+    data = json.loads(response.text)
+    return data
+
+
+def _send_request(endpoint, request_type, post_data=None, post_files=None):
+    """
+    Helper function for get and post that sends the request and handles any errors
+    that arise
+
+    Args:
+        Endpoint: the name of the endpoint to get data from
+        Request_type: Either post or get
+        Post_data: Any data to put in the post request's data field
+        Post_files: Any files to put in the post request's files field
+
+    Returns:
+        Response: A Response object generated from the request
+
+    Raises:
+        ClickException: Raised if cannot connect to host, permission issue
+        exists, user has not set a required field, or other request failure
+
+    """
+    host = get_config_value("host")
+    url = host + f"/api/v1/{endpoint}"
+    try:
+        if endpoint == "api-token-auth" and request_type == "post":
+            response = requests.post(url, data=post_data)
+        else:
+            token = get_config_value("token")
+            if endpoint == "teams" and request_type == "get":
+                headers = {"Authorization": f"Token {token}"}
+                response = requests.get(url, headers=headers)
+            else:
+                try:
+                    environment_id = get_config_value("current_environment_id")
+                except click.ClickException:
+                    raise click.ClickException(
+                        "Please set an active environment id using 'environment set'"
+                    )
+                headers = {
+                    "Authorization": f"Token {token}",
+                    "X-Environment-ID": f"{environment_id}",
```

Comment:
Similar comment to the one about the authorization token.  The API response will inform you failed to send the `X-Environment-Id` header in a situation that required it.  So you again should be able to just set the header if you have a value, not set it if you don't, do the request, and then if you get the 400 response telling you that the header is missing, raise your `ClickException("Please set an active environment id using 'environment set'")`.

The idea is to leverage the logic built into the API as best you can, rather than having to write a bunch of endpoint specific checks here in the CLI.